src/main/scala/org/constellation/gossip/GossipService.scala

```diff
@@ -1,47 +1,127 @@
+    */
+  protected def spreadFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Unit]
+
+  /**
+    * How to check if message was received by peer
+    */
+  protected def validationFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Boolean]
+
+  // TODO: Check if bisect is efficient enough for error handling Gossip
+  def recover(message: GossipMessage[A]): F[Unit] = {
+    val sequence = message.path.toIndexedSeq
 
-  protected def spread(data: A, fanout: Int)(spreadFn: (PeerClientMetadata, GossipMessage[A]) => F[Unit]): F[Unit] =
     for {
-      messages <- peerSampling
-        .selectPaths(fanout)
-        .map(_.map(GossipPath(_)).map(GossipMessage(data, _)))
+      _ <- messageTracker.fail(message.path.id)
+      failureNode <- bisectA(
+        (id: Id) => getClientMetadata(id).flatMap(validationFn(_, message))
+      )(sequence)
+      sequenceToRetry = sequence.dropWhile(failureNode.contains)
+      clients <- sequenceToRetry.toList.traverse(getClientMetadata)
+      _ <- clients.traverse(spreadFn(_, message)) // TODO: Force trigger peer healthcheck if it fails
+      _ <- messageTracker.remove(message.path.id)
+    } yield ()
+  }
 
-      _ <- messages.toList.traverse(spreadToNext(_)(spreadFn))
+  /**
+    * Start spreading the message to fanout
+    */
+  def spread(data: A): F[Unit] =
+    for {
+      _ <- logger.debug(s"Starting spreading of data: $data")
+      messages <- peerSampling.selectPaths
+        .map(_.map(GossipMessage(data, _)))
+      _ <- messages.map { msg =>
+        for {
+          _ <- messageTracker.start(msg)
+          _ <- logger.debug(s"Started path ${msg.path.id}: ${msg.path.toIndexedSeq.map(_.short).mkString(" -> ")}")
+          _ <- (for {
+            _ <- spreadToNext(msg)
+            _ <- T.sleep(getPathTimeout(msg.path))
+            succeeded <- messageTracker.isSuccess(msg)
+            _ <- if (succeeded) {
+              messageTracker.remove(msg.path.id) >> logger.debug(s"Succeeded path ${msg.path.id}")
+            } else {
+              recover(msg)
+            }
+          } yield ()).handleErrorWith(_ => recover(msg))
+        } yield ()
+      }.parSequence
     } yield ()
 
-  protected def spread(
-    message: GossipMessage[A]
-  )(spreadFn: (PeerClientMetadata, GossipMessage[A]) => F[Unit]): F[Unit] = {
-    val couldReceive = message.path.isNext(selfId)
-    if (!couldReceive) {
-      F.raiseError(IncorrectNextNodeOnPath(selfId.some, message.path.next))
-    } else {
-      spreadToNext(GossipMessage(message.data, message.path.accept))(spreadFn)
+  def validate(message: GossipMessage[A], senderId: Id): Either[GossipError, GossipMessage[A]] =
+    if (!couldSend(message, senderId)) {
+      Left(IncorrectSenderId(senderId))
+    } else if (!couldReceive(message)) {
+      Left(IncorrectReceiverId(selfId.some, message.path.next))
+    } else if (isEndOfCycle(message)) {
+      Left(EndOfCycle)
+    } else Right(message)
```

Comment:
please consider using pipeline for that .orElse .orElse or sth like that

Reply:
I removed that and implemented from the scratch as `Validated[Error, GossipMessage]` so let's ignore that for now.